roles/translator/src/proxy/bridge.rs

```diff
@@ -104,7 +142,183 @@ impl Bridge {
             future_jobs: vec![],
             last_p_hash: None,
             target,
-        }
+            first_ph_received: false,
+            request_ids: Id::new(),
+            pool_output_is_set: false,
+            solution_sender,
+        }));
+        match upstream_kind {
+            UpstreamKind::Standard => (),
+            UpstreamKind::WithNegotiator {
+                recv_tp,
+                recv_ph,
+                recv_coinbase_out,
+                send_mining_job,
+                ..
+            } => {
+                // open a channel so that jobs are created and last notify is updated also if no
+                // dowsntream connected
+                self_
+                    .safe_lock(|s| {
+                        s.channel_factory.new_extended_channel(0, 1.0, 0);
+                    })
+                    .unwrap();
+                Self::start_receiving_pool_coinbase_outs(self_.clone(), recv_coinbase_out);
+                Self::start_receiving_new_template(self_.clone(), recv_tp, send_mining_job.clone());
+                Self::start_receiving_new_prev_hash(self_.clone(), recv_ph, send_mining_job);
+            }
+        };
+        self_
+    }
+
+    pub fn start_receiving_pool_coinbase_outs(
+        self_mutex: Arc<Mutex<Self>>,
+        recv: Receiver<(Vec<TxOut>, u64)>,
+    ) {
+        task::spawn(async move {
+            while let Ok((outs, _id)) = recv.recv().await {
+                // TODO MVP3 assuming that only one coinbase is negotiated with the pool not handling
+                // different tokens, in order to do that we can use the out hashmap:
+                // self.tx_outs.insert(id, outs)
+                self_mutex
+                    .safe_lock(|s| {
+                        s.channel_factory.update_pool_outputs(outs);
+                        s.pool_output_is_set = true;
+                    })
+                    .unwrap();
+            }
+        });
+    }
+
+    pub fn start_receiving_new_template(
+        self_mutex: Arc<Mutex<Self>>,
+        new_template_reciver: Receiver<(NewTemplate<'static>, u64)>,
+        send_mining_job: Sender<SetCustomMiningJob<'static>>,
+    ) {
+        task::spawn(async move {
+            while !self_mutex
+                .safe_lock(|s| (s.first_ph_received && s.pool_output_is_set))
+                .unwrap()
+            {
+                tokio::task::yield_now().await;
```

Comment:
I'd recommend adding a debug! statement here in case this never exits - i don't think we'd ever know why.